src/refdb_fs.c

```diff
@@ -1479,12 +1489,25 @@ static int refdb_fs_backend__rename(
 	if ((error = loose_commit(&new_lock, new)) < 0)
 		goto cleanup;
 
+	/* We're done renaming refs, but we might have to update HEAD */
+	if (is_head) {
+		git_filebuf head_lock = GIT_FILEBUF_INIT;
+		head->target.symbolic = git__strdup(new_name);
+		if ((error = refdb_fs_backend__write_tail(&head_lock, _backend, head,
+				NULL, old_head_target, GIT_REFDB__FORCE_WRITE, who, message)) < 0) {
+			goto cleanup;
+		}
+		if ((error = loose_commit(&head_lock, head)) < 0)
+			goto cleanup;
+	}
```

Comment:
This code feels a bit like a layering violation to me. What happens in case where we rename a reference that is _not_ part of the FS refdb, but part of a different one? Or in case where HEAD itself is not part of the FS refdb? Sure, especially the second scenario is unlikely, but it just feels a bit weird to handle HEAD at this level.

What confuses me is that in a later commit, you delete code that is supposed to update HEAD. Why is it necessary to move the code into refdb_fs and why didn't the other code suffice? Furthermore, shouldn't these two commits belong together then, ideally with your line of thought why the move is necessary?

Reply:
I agree on the layering, though I'm not sure I understand your cases : the first one should not happen, because the current refdb is being called on behalf of it being the ref's backend. The second is already true because the refdb already relookups HEAD when writing.

The strange-two step process is there to make the reflog actually correct (before the last step, the `git_repository_set_head` call would insert a third, duplicate message). Looking back, they can be merged, so I don't know why I kept them separate. As to the why, it's because the reference renaming process is much more finicky with locks than what can be done at this level, so it's either "dump everything on the refdb so it can handle it" (this approach) or "make it more transaction-y" so we can work at an higher level.